### DISCUSSION TITLE
Best place to write custom directive implementation?

### DISCUSSION THREAD

**Discussion by taneba** [dotansimha/envelop#785] [Q&A] [answered]
Thank you all for awesome work, I have a question about writing custom directive.

When writing my own envelop plugin, it seems we could write custom directive logic in onExectute.onResolverCalled and onSchemaChange as well.
Which is the preferred place to write?

**Comment by n1ru4l** [comment, marked as answer]
@taneba Could you share what kind of directive you try to build? Depending on that the answer might differ, there is probably no "one suits all" solution.

**Reply by taneba** [reply]
I'm implementing `@isOwner` directive which will validate if decorated type or field is queried by it's owner and throw error if it's not.

```
directive @isOwner(ownerField: String) on FIELD_DEFINITION | OBJECT

type User {
   id: ID!
   email: String! @isOwner(ownerField: "id")
   name: String
}

type Post @isOwner(ownerField: "authorId") {
  authorId: String!
  ...
}
```

I'm doing in onSchemaChange now ([code](https://github.com/taneba/fullstack-graphql-app/blob/main/backend/src/utils/useOwnerCheck.ts)), but I'm not sure if i'm doing in a preferable way.

**Reply by n1ru4l** [reply]
For this use-case that looks totally fine with me and `onSchemaChange` seems like the right place to do it! In case your used schema is changed frequently you could furthermore utilize a WeakMap for caching the schema transform.

**Reply by taneba** [reply]
@n1ru4l 
Thanks, good to hear that! Is it because it needs schema override which can not be done in onExecute, right?

> In case your used schema is changed frequently you could furthermore utilize a WeakMap for caching the schema transform

Yeah, I've learned that trick from useResponseCache plugin and I'm using it :)

**Reply by n1ru4l** [reply]
@taneba I think this could be a great example for our documentation - Would you mind sending a PR with a short guide on how to add custom schema directive logic to your schema with envelop?

**Reply by taneba** [reply]
@n1ru4l 
Of course I'd love to! IMO isOwner directive is a bit complex, so how about uppercase directive, which is usually shown in a custom directive example?

**Reply by n1ru4l** [reply]
IMHO an `uppercase` directive should instead be a field argument 😓 . It is really hard to find legit use-cases where a schema directive actually makes sense 🤔
